doc/arch/adr-002-digest-mvp.md

```diff
@@ -0,0 +1,187 @@
+- Ensure that multiple digests for the same period are not run concurrently
+
+##### Should
+
+- Be configurable as to what times a digest runs from and until
+
+#### `DigestRun`
+
+A model which represents a distinct run of a period of digest (daily or
+weekly).
+
+Has a responsibility to persist data that is related to a full digest run for
+all subscribers.
+
+##### Shall
+
+- Store information such as
+  - Date digest is run on
+  - The period the digest is for
+  - The start time for the digest period
+  - The end time for the digest period
+
+##### Should
+
+- Store a timestamp for when all the emails were created for the digest, to act
+  as indicator it is complete.
+
+##### May
+
+- Have a unique index for digest date and period, however this would limit
+  ability to re-run a digest in the case of a problem.
+
+#### `DigestSubscriberQuery`
+
+Responsible for taking a [`DigestRun`](#digestrun) instance and using that to
+determine which subscribers should receive an email for the digest period
+
+##### Shall
+
+- Use a `DigestRun` to determine which subscribers should receive an email for
+  a digest period
+
+##### Should
+
+- Return a list of ids of subscribers
+
+##### May
+
+- Return an ActiveRecord scope to allow the caller to deal with this returning
+  an unlimited output.
+
+#### `DigestRunSubscriber`
+
+A model which is used to associate a [`DigestRun`](#digestrun) with a subscriber
+instance. Will mostly be used as a persisted piece of data that logs which
+subscribers receive an email due to a digest. It is unlikely to be needed
+long term.
+
+##### Shall
+
+- Store
+  - An association to `DigestRun`
+  - An association to `Subscriber`
+
+##### Should
+
+- Store a timestamp for when it has been completed, can be used therefore to
+  work out if a `DigestRun` is complete or not.
+
+#### `DigestGenerationWorker`
+
+A worker entity that accepts the argument of a
+[`DigestRunSubscriber`](#digestrunsubscriber) and has the responsibility to
+create an Email entity for the associated subscriber.
+
+##### Shall
+
+- Interface with [`ContentChangeQuery`](#contentchangequery) to load content
+  changes
+- Interface with [`DigestEmailRenderer`](#digestemailrenderer) to render email
+  content
+- Create an `Email` entity
+- Create a `SubscriptionContent` entity, which will require knowing which
+  subscription an email is for
+
+##### Should
+
+- Mark a `DigestRunSubscriber` as completed
+- Mark `DigestRun` as completed if all `DigestRunSubscriber` are complete for
+  the `DigestRun`
+
+##### May
+
+- Liaise with a `DigestGeneratorService` to do the work if the code for the
+  worker becomes complicated or should be more re-usable
+
+#### `ContentChangeQuery`
+
+For a given [`DigestRunSubscriber`](#digestrunsubscriber) this determines which,
+if any, `ContentChange` entities apply and will make up the contents for the
+Email.
+
+##### Should
+
+- Return which subscriptions are associated with a `ContentChange`. It may be
```

Comment:
Do you mean this would return the `ContentChanges` rather than the subscriptions?

Reply:
I thought we'd want to know both but of course if we adapt it to accept subscription it wouldn't need this.